timers: do not retain a reference to the async store after firing

mcollina opened the pull request:
After firing timers, we can clean them up by iterating over all active stores and setting the relevant symbols to undefined.

This is still a draft because we will need to extend it to immediates and intervals, too.

Fixes #53408

mcollina commented:
I've opened a draft PR to discuss a possible solution to https://github.com/nodejs/node/issues/53408.

cc @nodejs/diagnostics @nodejs/timers let me know what you think.

Flarna commented on lib/internal/timers.js:
```diff
@@ -429,6 +432,20 @@ function setPosition(node, pos) {
   node.priorityQueuePosition = pos;
 }
 
+function removeAllStores (timer) {
```
What about moving this function into async_hooks to avoid the need to expose `getActiveStores()` there?
An alternative would be to use a single object on the resource which holds all stores. This would reduce the number of properties added on resource objects - and avoid the need of a for loop to clear it.

Qard replied:
A helper that receives a resource and clears all the store data off it would be better than exposing the stores list as we do now. As it is presently, one could get and retain access to a store in a way unintentional from the perspective of the store owner, which is a bit dangerous from a memory perspective.

Flarna commented on lib/internal/timers.js:
```diff
@@ -594,6 +611,9 @@ function getTimerCallbacks(runNextTicks) {
             timeoutInfo[0]--;
 
+          removeAllStores(timer);
+          timer._onTimeout = undefined;
```
maybe clear also `_timerArgs`.

mcollina commented:
I've updated this PR to address the review feedback:

1. Instead of exposing `getActiveStores()`, added a `clearStores(resource)` helper function in `async_hooks` that takes a resource and clears all store references from it. This avoids exposing the internal stores list.
2. Added cleanup for `_timerArgs` in addition to `_onTimeout` and async stores.
3. Extended the cleanup to immediates (`setImmediate`) and intervals (`setInterval`) as well.
4. Updated the test to cover all three cases.

cc @Flarna @Qard

Flarna commented on lib/internal/async_local_storage/async_hooks.js:
```diff
@@ -151,3 +157,4 @@ class AsyncLocalStorage {
 }
 
 module.exports = AsyncLocalStorage;
+module.exports.clearStores = clearStores;
```
Doesn't this add a static `AsyncLocalStorage#clearStores`  function visible to all ALS users so a new public API?

Flarna commented on lib/internal/timers.js:
```diff
@@ -577,6 +585,10 @@ function getTimerCallbacks(runNextTicks) {
           timer._destroyed = true;
 
+          removeStoresFromResource(timer);
+          timer._onTimeout = undefined;
```
Maybe move into a `cleanTimer(timer)` function as these two lines are there twice in the file.

Flarna commented on lib/async_hooks.js:
```diff
@@ -279,6 +279,15 @@ class AsyncResource {
+    // so there is nothing to clear.
+    return;
+  }
+  require('internal/async_local_storage/async_hooks').clearStores(resource);
```
Maybe worth to convert this into a lazy load once instead calling `require` always.

Flarna commented on lib/internal/timers.js:
```diff
@@ -125,6 +131,36 @@ const AsyncContextFrame = require('internal/async_context_frame');
 
 const async_context_frame = Symbol('kAsyncContextFrame');
 
+function removeStoresFromResource(resource) {
```
I think this should be extended to somthing like this to cover also the async context frame variant of ALS:
```js
function removeStoresFromResource(resource) {
  if (AsyncContextFrame.enabled) {
    if (resource[async_context_frame] !== undefined) {
      resource[async_context_frame] = undefined;
    }
  } else {
    if (resource[async_local_storage_context_symbol] !== undefined) {
      resource[async_local_storage_context_symbol] = undefined;
    }
  }
}
```

Flarna commented on test/parallel/test-timers-async-store-leak.js:
```diff
@@ -0,0 +1,109 @@
+// Flags: --expose-internals --no-async-context-frame
```
I think this should be tested with and without async-context-frame

And a test via GC would be likely helpful, e.g. somthing like this:
```js
{
  const asyncLocalStorage = new AsyncLocalStorage();
  let store = {};
  let timeout;
  const weakStore = new WeakRef(store);
  asyncLocalStorage.run(store, common.mustCall(() => {
    timeout = setTimeout(common.mustNotCall(), common.platformTimeout(1000));
    assert.strictEqual(weakStore.deref(), store);
    clearTimeout(timeout);
  }));
  store = null;
  setImmediate(() => {
    global.gc();
    assert.strictEqual(weakStore.deref(), undefined);
  });
}
```

Flarna commented:
@mcollina seems you forced push some old commit or a mixture of old and new. At least it looks to me like parts of the recent changes got lost.

Any chance to avoid forced push to ease reviews?

mcollina commented:
I had to force-push for rebase.

Flarna commented on lib/timers.js:
```diff
@@ -239,6 +247,8 @@ function clearImmediate(immediate) {
   emitDestroy(immediate[async_id_symbol]);
 
   immediate._onImmediate = null;
+  immediate._argv = undefined;
```
This line and that one above shall be removed as `cleanImmediate()` does this.

Flarna commented on lib/internal/timers.js:
```diff
@@ -698,8 +763,11 @@ module.exports = {
   kTimeout: Symbol('timeout'), // For hiding Timeouts on other internals.
   async_id_symbol,
   trigger_async_id_symbol,
+  async_context_frame,
```
ist this export needed?

mcollina replied:
Yes https://github.com/nodejs/node/pull/53443/changes#diff-48776d27fdcc917f21151512814fd67e24cace709ce6e18014623ed7d7fd930dR30

Flarna replied:
ah ok, for tests only.

mcollina commented:
@Flarna PTAL

Flarna commented:
I still don't understand why rebase+squash is done again and again in this PR.
I get that a rebase is needed every now and then, in special for that long running PRs.
But I see no need to squash fixup commits at the same time.

mcollina commented:
> I still don't understand why rebase+squash is done again and again in this PR.
I get that a rebase is needed every now and then, in special for that long running PRs.
But I see no need to squash fixup commits at the same time.

Sorry, I got a habit of squashing the review fixes when rebasing, mostly because other collaborators told me they preferred that way.
